### ISSUE TITLE
@induct breaks termination

### ISSUE THREAD

**Issue by Gorzen** [epfl-lara/stainless#644]
If we check the termination for the following program we get a successful output.
```scala
import stainless.lang._
import stainless.annotation._
import stainless.proof._

object BugTermination {
  abstract class SemiGroup[T] {
    def combine(x: T, y: T): T

    @law
    def law_associativity(x: T, y: T, z: T): Boolean = {
      combine(combine(x, y), z) == combine(x, combine(y, z))
    }
  }
  
  case class BugTermination[T](ev: SemiGroup[T]) extends SemiGroup[Option[T]] {
    def combine(x: Option[T], y: Option[T]): Option[T] = {
      (x, y) match {
        case (Some(a), Some(b)) => Some(ev.combine(a, b))
        case (Some(a), None()) => Some(a)
        case (None(), Some(b)) => Some(b)
        case (None(), None()) => None[T]()
      }
    }
  
    override def law_associativity(x: Option[T], y: Option[T], z: Option[T]): Boolean = {
      (combine(combine(x, y), z) == combine(x, combine(y, z))) because {
        (x, y, z) match {
          case (Some(a), Some(b), Some(c)) => check(ev.law_associativity(a, b, c))
          case _ => true
        }
      }
    }
  }
}
```
*The output...*
```
[info] [  Info  ] Result for isSemiGroupExt
[info] [  Info  ]  => CLEARED (Non-recursive)
[info] [  Info  ] Result for combine
[info] [  Info  ]  => CLEARED (Non-recursive)
[info] [  Info  ] Result for isOption
[info] [  Info  ]  => CLEARED (Non-recursive)
[info] [  Info  ] Result for combine
[info] [  Info  ]  => CLEARED (Non-recursive)
[info] [  Info  ] Result for combine
[info] [  Info  ]  => CLEARED (Non-recursive)
[info] [  Info  ] Result for combine
[info] [  Info  ]  => CLEARED (Non-recursive)
[info] [  Info  ] Result for isBugTermination
[info] [  Info  ]  => CLEARED (Unchecked)
[info] [  Info  ] Result for isSemiGroup
[info] [  Info  ]  => CLEARED (Unchecked)
[info] [  Info  ] Result for combine
[info] [  Info  ]  => CLEARED (Relation Processor comparing argument lists lexicographically)
[info] [  Info  ] Result for combine
[info] [  Info  ]  => CLEARED (Relation Processor comparing argument lists lexicographically)
[info] [  Info  ] Result for law_associativity
[info] [  Info  ]  => CLEARED (Non-recursive)
[info] [  Info  ] Result for law_associativity
[info] [  Info  ]  => CLEARED (Non-recursive)
[info] [  Info  ] Result for isSemiGroupExt
[info] [  Info  ]  => CLEARED (Non-recursive)
[info] [  Info  ] Result for combine
[info] [  Info  ]  => CLEARED (Non-recursive)
[info] [  Info  ] Result for isOption
[info] [  Info  ]  => CLEARED (Non-recursive)
[info] [  Info  ] Result for combine
[info] [  Info  ]  => CLEARED (Non-recursive)
[info] [  Info  ] Result for law_associativity
[info] [  Info  ]  => CLEARED (Non-recursive)
[info] [  Info  ] Result for combine
[info] [  Info  ]  => CLEARED (Non-recursive)
[info] [  Info  ] Result for combine
[info] [  Info  ]  => CLEARED (Non-recursive)
[info] [  Info  ] Result for isBugTermination
[info] [  Info  ]  => CLEARED (Unchecked)
[info] [  Info  ] Result for isSemiGroup
[info] [  Info  ]  => CLEARED (Unchecked)
[info] [  Info  ] Result for law_associativity
[info] [  Info  ]  => CLEARED (Non-recursive)
[info] [  Info  ] Result for law_associativity
[info] [  Info  ]  => CLEARED (Non-recursive)
[info] [  Info  ] Result for law_associativity
[info] [  Info  ]  => CLEARED (Non-recursive)
[info] [  Info  ] Result for isSemiGroupExt
[info] [  Info  ]  => CLEARED (Non-recursive)
[info] [  Info  ] Result for combine
[info] [  Info  ]  => CLEARED (Non-recursive)
[info] [  Info  ] Result for isOption
[info] [  Info  ]  => CLEARED (Non-recursive)
[info] [  Info  ] Result for law_associativity
[info] [  Info  ]  => CLEARED (Non-recursive)
[info] [  Info  ] Result for law_associativity
[info] [  Info  ]  => CLEARED (Non-recursive)
[info] [  Info  ] Result for because
[info] [  Info  ]  => CLEARED (Non-recursive)
[info] [  Info  ] Result for check
[info] [  Info  ]  => CLEARED (Non-recursive)
[info] [  Info  ] Result for combine
[info] [  Info  ]  => CLEARED (Non-recursive)
[info] [  Info  ] Result for law_associativity
[info] [  Info  ]  => CLEARED (Non-recursive)
[info] [  Info  ] Result for combine
[info] [  Info  ]  => CLEARED (Non-recursive)
[info] [  Info  ] Result for combine
[info] [  Info  ]  => CLEARED (Non-recursive)
[info] [  Info  ] Result for isBugTermination
[info] [  Info  ]  => CLEARED (Unchecked)
[info] [  Info  ] Result for isSemiGroup
[info] [  Info  ]  => CLEARED (Unchecked)
[info] [  Info  ] Result for law_associativity
[info] [  Info  ]  => CLEARED (Relation Processor comparing argument lists lexicographically)
[info] [  Info  ] Result for law_associativity
[info] [  Info  ]  => CLEARED (Relation Processor comparing argument lists lexicographically)
[info] [  Info  ] Result for isSemiGroupExt
[info] [  Info  ]  => CLEARED (Non-recursive)
[info] [  Info  ] Result for combine
[info] [  Info  ]  => CLEARED (Non-recursive)
[info] [  Info  ] Result for isOption
[info] [  Info  ]  => CLEARED (Non-recursive)
[info] [  Info  ] Result for law_associativity
[info] [  Info  ]  => CLEARED (Non-recursive)
[info] [  Info  ] Result for combine
[info] [  Info  ]  => CLEARED (Non-recursive)
[info] [  Info  ] Result for law_associativity
[info] [  Info  ]  => CLEARED (Non-recursive)
[info] [  Info  ] Result for combine
[info] [  Info  ]  => CLEARED (Non-recursive)
[info] [  Info  ] Result for combine
[info] [  Info  ]  => CLEARED (Non-recursive)
[info] [  Info  ] Result for isBugTermination
[info] [  Info  ]  => CLEARED (Unchecked)
[info] [  Info  ] Result for isSemiGroup
[info] [  Info  ]  => CLEARED (Unchecked)
[info] [  Info  ]   ┌───────────────────┐
[info] [  Info  ] ╔═╡ stainless summary ╞═════════════════════════════════════════════════════════════════════════╗
[info] [  Info  ] ║ └───────────────────┘                                                                         ║
[info] [  Info  ] ║ combine$8             ✓ terminates  ../../BugTermination/BugTerminationLong.scala:7:5   2.525 ║
[info] [  Info  ] ║ combine$9             ✓ terminates  ../../BugTermination/BugTerminationLong.scala:16:5  0.000 ║
[info] [  Info  ] ║ combine$10            ✓ terminates  ../../BugTermination/BugTerminationLong.scala:7:5   0.000 ║
[info] [  Info  ] ║ combine$11            ✓ terminates  ../../BugTermination/BugTerminationLong.scala:7:5   0.000 ║
[info] [  Info  ] ║ combine$12            ✓ terminates  ../../BugTermination/BugTerminationLong.scala:16:5  0.000 ║
[info] [  Info  ] ║ combine$13            ✓ terminates  ../../BugTermination/BugTerminationLong.scala:7:5   0.000 ║
[info] [  Info  ] ║ law_associativity$4   ✓ terminates  ../../BugTermination/BugTerminationLong.scala:10:5  3.546 ║
[info] [  Info  ] ║ law_associativity$5   ✓ terminates  ../../BugTermination/BugTerminationLong.scala:25:5  0.000 ║
[info] [  Info  ] ║ law_associativity$6   ✓ terminates  ../../BugTermination/BugTerminationLong.scala:10:5  0.002 ║
[info] [  Info  ] ║ law_associativity$7   ✓ terminates  ../../BugTermination/BugTerminationLong.scala:10:5  0.004 ║
[info] [  Info  ] ║ law_associativity$8   ✓ terminates  ../../BugTermination/BugTerminationLong.scala:10:5  0.000 ║
[info] [  Info  ] ║ law_associativity$9   ✓ terminates  ../../BugTermination/BugTerminationLong.scala:25:5  0.000 ║
[info] [  Info  ] ║ law_associativity$10  ✓ terminates  ../../BugTermination/BugTerminationLong.scala:10:5  0.000 ║
[info] [  Info  ] ║ law_associativity$11  ✓ terminates  ../../BugTermination/BugTerminationLong.scala:10:5  0.000 ║
[info] [  Info  ] ╟┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄╢
[info] [  Info  ] ║ total: 14   valid: 14   (0 from cache) invalid: 0    unknown: 0    time:   6.077              ║
[info] [  Info  ] ╚═══════════════════════════════════════════════════════════════════════════════════════════════╝

```
However, if we add an induct flag, the termination now breaks. I.e. by simply changing:
```scala
override def law_associativity(x: Option[T], y: Option[T], z: Option[T]): Boolean = {
```
to
```scala
override def law_associativity(@induct x: Option[T], y: Option[T], z: Option[T]): Boolean = {
```
we break it.
```scala
import stainless.lang._
import stainless.annotation._
import stainless.proof._

object BugTermination {
  abstract class SemiGroup[T] {
    def combine(x: T, y: T): T

    @law
    def law_associativity(x: T, y: T, z: T): Boolean = {
      combine(combine(x, y), z) == combine(x, combine(y, z))
    }
  }
  
  case class BugTermination[T](ev: SemiGroup[T]) extends SemiGroup[Option[T]] {
    def combine(x: Option[T], y: Option[T]): Option[T] = {
      (x, y) match {
        case (Some(a), Some(b)) => Some(ev.combine(a, b))
        case (Some(a), None()) => Some(a)
        case (None(), Some(b)) => Some(b)
        case (None(), None()) => None[T]()
      }
    }
  
    override def law_associativity(@induct x: Option[T], y: Option[T], z: Option[T]): Boolean = {
      (combine(combine(x, y), z) == combine(x, combine(y, z))) because {
        (x, y, z) match {
          case (Some(a), Some(b), Some(c)) => check(ev.law_associativity(a, b, c))
          case _ => true
        }
      }
    }
  }
}

```
*The output...*
```
[info] [  Info  ] Result for isSemiGroupExt
[info] [  Info  ]  => CLEARED (Non-recursive)
[info] [  Info  ] Result for combine
[info] [  Info  ]  => CLEARED (Non-recursive)
[info] [  Info  ] Result for isOption
[info] [  Info  ]  => CLEARED (Non-recursive)
[info] [  Info  ] Result for combine
[info] [  Info  ]  => CLEARED (Non-recursive)
[info] [  Info  ] Result for combine
[info] [  Info  ]  => CLEARED (Non-recursive)
[info] [  Info  ] Result for combine
[info] [  Info  ]  => CLEARED (Non-recursive)
[info] [  Info  ] Result for isBugTermination
[info] [  Info  ]  => CLEARED (Unchecked)
[info] [  Info  ] Result for isSemiGroup
[info] [  Info  ]  => CLEARED (Unchecked)
[info] [  Info  ] Result for combine
[info] [  Info  ]  => CLEARED (Relation Processor comparing argument lists lexicographically)
[info] [  Info  ] Result for combine
[info] [  Info  ]  => CLEARED (Relation Processor comparing argument lists lexicographically)
[info] [  Info  ] Result for law_associativity
[info] [  Info  ]  => CLEARED (Non-recursive)
[info] [  Info  ] Result for law_associativity
[info] [  Info  ]  => CLEARED (Non-recursive)
[info] [  Info  ] Result for isSemiGroupExt
[info] [  Info  ]  => CLEARED (Non-recursive)
[info] [  Info  ] Result for combine
[info] [  Info  ]  => CLEARED (Non-recursive)
[info] [  Info  ] Result for isOption
[info] [  Info  ]  => CLEARED (Non-recursive)
[info] [  Info  ] Result for combine
[info] [  Info  ]  => CLEARED (Non-recursive)
[info] [  Info  ] Result for law_associativity
[info] [  Info  ]  => CLEARED (Non-recursive)
[info] [  Info  ] Result for combine
[info] [  Info  ]  => CLEARED (Non-recursive)
[info] [  Info  ] Result for combine
[info] [  Info  ]  => CLEARED (Non-recursive)
[info] [  Info  ] Result for isBugTermination
[info] [  Info  ]  => CLEARED (Unchecked)
[info] [  Info  ] Result for isSemiGroup
[info] [  Info  ]  => CLEARED (Unchecked)
[info] [  Info  ] Result for law_associativity
[info] [  Info  ]  => CLEARED (Non-recursive)
[info] [  Info  ] Result for law_associativity
[info] [  Info  ]  => CLEARED (Non-recursive)
[info] [  Info  ] Result for law_associativity
[info] [  Info  ]  => CLEARED (Non-recursive)
[info] [  Info  ] Result for isSemiGroupExt
[info] [  Info  ]  => CLEARED (Non-recursive)
[info] [  Info  ] Result for combine
[info] [  Info  ]  => CLEARED (Non-recursive)
[info] [  Info  ] Result for isOption
[info] [  Info  ]  => CLEARED (Non-recursive)
[info] [  Info  ] Result for law_associativity
[info] [  Info  ]  => CLEARED (Non-recursive)
[info] [  Info  ] Result for law_associativity
[info] [  Info  ]  => CLEARED (Non-recursive)
[info] [  Info  ] Result for because
[info] [  Info  ]  => CLEARED (Non-recursive)
[info] [  Info  ] Result for check
[info] [  Info  ]  => CLEARED (Non-recursive)
[info] [  Info  ] Result for OptionPrimitiveSize
[info] [  Info  ]  => CLEARED (Non-recursive)
[info] [  Info  ] Result for combine
[info] [  Info  ]  => CLEARED (Non-recursive)
[info] [  Info  ] Result for law_associativity
[info] [  Info  ]  => CLEARED (Non-recursive)
[info] [  Info  ] Result for combine
[info] [  Info  ]  => CLEARED (Non-recursive)
[info] [  Info  ] Result for combine
[info] [  Info  ]  => CLEARED (Non-recursive)
[info] [  Info  ] Result for isBugTermination
[info] [  Info  ]  => CLEARED (Unchecked)
[info] [  Info  ] Result for isSemiGroup
[info] [  Info  ]  => CLEARED (Unchecked)
[info] [Warning ] ==> UNKNOWN: measure cannot be shown to decrease for (transitive) call law_associativity$5((x$295: Object$0) => true, @unchecked {
[info] [Warning ]   val x$298: { x$296: Object$0 | @unchecked isSemiGroup$0(x$296, T$1626) } = thiss$527
[info] [Warning ]   assert(isBugTermination$0(x$298, (x$301: Object$0) => true), "Cast error")
[info] [Warning ]   x$298
[info] [Warning ] }, {
[info] [Warning ]   assume(x$2497 is Option$3 && isOption$0[Object$0](@unchecked x$2497.value$14, (x$303: Object$0) => true))
[info] [Warning ]   @unchecked {
[info] [Warning ]     val x$305: { x$304: Object$0 | @unchecked T$1626(x$304) } = x$2497
[info] [Warning ]     assert(x$305 is Option$3 && isOption$0[Object$0](@unchecked x$305.value$14, (x$307: Object$0) => true), "Cast error")
[info] [Warning ]     @unchecked x$305.value$14
[info] [Warning ]   }
[info] [Warning ] }, {
[info] [Warning ]   assume(y$643 is Option$3 && isOption$0[Object$0](@unchecked y$643.value$14, (x$309: Object$0) => true))
[info] [Warning ]   @unchecked {
[info] [Warning ]     val x$311: { x$310: Object$0 | @unchecked T$1626(x$310) } = y$643
[info] [Warning ]     assert(x$311 is Option$3 && isOption$0[Object$0](@unchecked x$311.value$14, (x$313: Object$0) => true), "Cast error")
[info] [Warning ]     @unchecked x$311.value$14
[info] [Warning ]   }
[info] [Warning ] }, {
[info] [Warning ]   assume(z$189 is Option$3 && isOption$0[Object$0](@unchecked z$189.value$14, (x$315: Object$0) => true))
[info] [Warning ]   @unchecked {
[info] [Warning ]     val x$317: { x$316: Object$0 | @unchecked T$1626(x$316) } = z$189
[info] [Warning ]     assert(x$317 is Option$3 && isOption$0[Object$0](@unchecked x$317.value$14, (x$319: Object$0) => true), "Cast error")
[info] [Warning ]     @unchecked x$317.value$14
[info] [Warning ]   }
[info] [Warning ] })
[info] [Warning ] Result for law_associativity
[info] [Warning ]  => BROKEN (Decreases Processor)
[info] [Warning ]   Decreases check failed for law_associativity$4
[info] [Warning ] Result for law_associativity
[info] [Warning ]  => BROKEN (Decreases Processor)
[info] [Warning ]   Decreases check failed for law_associativity$5
[info] [  Info  ] Result for isSemiGroupExt
[info] [  Info  ]  => CLEARED (Non-recursive)
[info] [  Info  ] Result for combine
[info] [  Info  ]  => CLEARED (Non-recursive)
[info] [  Info  ] Result for isOption
[info] [  Info  ]  => CLEARED (Non-recursive)
[info] [  Info  ] Result for law_associativity
[info] [  Info  ]  => CLEARED (Non-recursive)
[info] [  Info  ] Result for combine
[info] [  Info  ]  => CLEARED (Non-recursive)
[info] [  Info  ] Result for law_associativity
[info] [  Info  ]  => CLEARED (Non-recursive)
[info] [  Info  ] Result for combine
[info] [  Info  ]  => CLEARED (Non-recursive)
[info] [  Info  ] Result for combine
[info] [  Info  ]  => CLEARED (Non-recursive)
[info] [  Info  ] Result for isBugTermination
[info] [  Info  ]  => CLEARED (Unchecked)
[info] [  Info  ] Result for isSemiGroup
[info] [  Info  ]  => CLEARED (Unchecked)
[info] [  Info  ]   ┌───────────────────┐
[info] [  Info  ] ╔═╡ stainless summary ╞═════════════════════════════════════════════════════════════════════════════════════╗
[info] [  Info  ] ║ └───────────────────┘                                                                                     ║
[info] [  Info  ] ║ combine$8             ✓ terminates              ../../BugTermination/BugTerminationLong.scala:7:5   2.762 ║
[info] [  Info  ] ║ combine$9             ✓ terminates              ../../BugTermination/BugTerminationLong.scala:16:5  0.000 ║
[info] [  Info  ] ║ combine$10            ✓ terminates              ../../BugTermination/BugTerminationLong.scala:7:5   0.000 ║
[info] [  Info  ] ║ combine$11            ✓ terminates              ../../BugTermination/BugTerminationLong.scala:7:5   0.000 ║
[info] [  Info  ] ║ combine$12            ✓ terminates              ../../BugTermination/BugTerminationLong.scala:16:5  0.000 ║
[info] [  Info  ] ║ combine$13            ✓ terminates              ../../BugTermination/BugTerminationLong.scala:7:5   0.000 ║
[info] [  Info  ] ║ law_associativity$4   ✗ failed decreases check  ../../BugTermination/BugTerminationLong.scala:10:5  1.500 ║
[info] [  Info  ] ║ law_associativity$5   ✗ failed decreases check  ../../BugTermination/BugTerminationLong.scala:25:5  0.000 ║
[info] [  Info  ] ║ law_associativity$6   ✓ terminates              ../../BugTermination/BugTerminationLong.scala:10:5  0.002 ║
[info] [  Info  ] ║ law_associativity$7   ✓ terminates              ../../BugTermination/BugTerminationLong.scala:10:5  0.004 ║
[info] [  Info  ] ║ law_associativity$8   ✓ terminates              ../../BugTermination/BugTerminationLong.scala:10:5  0.000 ║
[info] [  Info  ] ║ law_associativity$9   ✓ terminates              ../../BugTermination/BugTerminationLong.scala:25:5  0.000 ║
[info] [  Info  ] ║ law_associativity$10  ✓ terminates              ../../BugTermination/BugTerminationLong.scala:10:5  0.000 ║
[info] [  Info  ] ║ law_associativity$11  ✓ terminates              ../../BugTermination/BugTerminationLong.scala:10:5  0.000 ║
[info] [  Info  ] ╟┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄╢
[info] [  Info  ] ║ total: 14   valid: 12   (0 from cache) invalid: 2    unknown: 0    time:   4.268                          ║
[info] [  Info  ] ╚═══════════════════════════════════════════════════════════════════════════════════════════════════════════╝
```

**Comment by Gorzen** [comment]
Fixed by #657

**Comment by romac** [comment]
Right, thanks!
